src/wp-includes/class-wp-query.php

```diff
@@ -3521,8 +3521,20 @@ public function get_posts() {
 			/** @var WP_Post[] */
 			$this->posts = array_map( 'get_post', $this->posts );
 
-			if ( $q['cache_results'] ) {
-				update_post_caches( $this->posts, $post_type, $q['update_post_term_cache'], $q['update_post_meta_cache'] );
+			if ( $q['cache_results'] && $id_query_is_cacheable ) {
+				update_post_cache( $this->posts );
+			}
+
+			$post_ids = wp_list_pluck( $this->posts, 'ID' );
+
+			if ( $q['update_post_meta_cache'] ) {
+				update_postmeta_cache( $post_ids );
+			}
+
+			if ( $q['update_post_term_cache'] ) {
+				$post_types = array_map( 'get_post_type', $this->posts );
+				$post_types = array_unique( $post_types );
+				update_object_term_cache( $post_ids, $post_types );
 			}
```

Comment:
Instead of duplicating all this logic from `update_post_caches()`, can we maybe find another solution where we change `update_post_caches()` (or the underlying `update_post_cache()`) instead? Maybe we can add logic there to only cache post objects if they have certain fields present?

That way we could keep this class untouched.

Reply:
I don't know how that would work reliably. The query can be filtered in multiple ways. We don't know what fields were returned in the query. Is the post_content empty in the database or was it filtered to be this way. For backward compatibility reasons,I don't want to mess with `update_post_caches`. 

To ensure that the same logic runs, which is why I copied the logic over. This change also fixes a long standing bug, where post meta / terms were not primed if `cache_results` is set to false but `update_post_meta_cache` / `update_post_term_cache` were set to true.

Reply:
@spacedmonkey That's fair. However, the logic you have here in lines 3543-3544 differs from what is in `update_post_caches()` slightly, and because overall that logic is somewhat duplicated, I think we can find a cleaner solution to avoid the duplicate code.

Here's what I would propose:
* Introduce a new function `update_post_term_caches( WP_Post[] $posts, string|array $post_type = 'post' )`.
* Move all logic from within the `$update_term_cache` clause in `update_post_caches()` into that function, then replace it in the original function with a call to the new `update_post_term_caches( $posts, $post_type )`.
* Then, in here we can also simply call the single function `update_post_term_caches( $this->posts, $post_type )` instead of duplicating logic to figure out the post types to pass.

This way we avoid all duplicate code, and the new function `update_post_term_cache()` becomes a post-specific wrapper for the more general `update_object_term_cache()`. Plus, it fits nicely with the other existing functions `update_post_author_caches()` and `update_post_parent_caches()`.